File: fe/packages/components/src/component/map/marker-motion.js
const radians = Math.PI / 180
const delta = value => ((value + 540) % 360) - 180

export function distance(a, b) {
	const dy = (b.latitude - a.latitude) * radians
	const dx = delta(b.longitude - a.longitude) * radians
	const h = Math.sin(dy / 2) ** 2 + Math.cos(a.latitude * radians) * Math.cos(b.latitude * radians) * Math.sin(dx / 2) ** 2
	return 6371008.8 * 2 * Math.asin(Math.sqrt(Math.min(1, h)))
}

// Precompute distance once; each frame does a binary search, not a route scan.
export function markerMotion(path, { angle = 0, rotate = angle, autoRotate = false, separateRotation = false } = {}) {
	const lengths = [0]
	for (let i = 1; i < path.length; i++) lengths.push(lengths[i - 1] + distance(path[i - 1], path[i]))
	const total = lengths.at(-1)
	return (progress) => {
		const rotation = separateRotation ? Math.min(1, progress * 2) : progress
		const movement = separateRotation ? Math.max(0, progress * 2 - 1) : progress
		const traveled = movement * total
		let low = 1; let high = path.length - 1
		while (low < high) { const mid = (low + high) >> 1; if (lengths[mid] < traveled) low = mid + 1; else high = mid }
		const a = path[low - 1]; const b = path[low]
		const fraction = lengths[low] === lengths[low - 1] ? movement : (traveled - lengths[low - 1]) / (lengths[low] - lengths[low - 1])
		const dx = delta(b.longitude - a.longitude) * radians
		const bearing = Math.atan2(Math.sin(dx) * Math.cos(b.latitude * radians),
			Math.cos(a.latitude * radians) * Math.sin(b.latitude * radians) - Math.sin(a.latitude * radians) * Math.cos(b.latitude * radians) * Math.cos(dx)) / radians
		return {
			longitude: movement === 1 ? b.longitude : delta(a.longitude + delta(b.longitude - a.longitude) * fraction),
			latitude: a.latitude + (b.latitude - a.latitude) * fraction,
			rotate: autoRotate && movement > 0 && total > 0 ? (bearing + 360) % 360 : angle + delta(rotate - angle) * rotation,
			distance: traveled,
		}
	}
}

File: android/map-amap/src/main/kotlin/com/didi/dimina/map/amap/MarkerMotion.kt
package com.didi.dimina.map.amap

import kotlin.math.*

internal data class MotionPoint(val longitude: Double, val latitude: Double)
internal data class MotionSample(val point: MotionPoint, val rotate: Double, val distance: Double)

/** Distance table is immutable; sampling costs O(log n), with no SDK or bridge calls. */
internal class MarkerMotion(private val path: List<MotionPoint>, private val angle: Double,
    private val rotate: Double, private val autoRotate: Boolean, private val separateRotation: Boolean) {
    private val lengths = DoubleArray(path.size)
    init {
        require(path.size >= 2) { "Path requires at least two points" }
        for (i in 1 until path.size) {
            val a = path[i - 1]; val b = path[i]
            val h = sin(Math.toRadians(b.latitude - a.latitude) / 2).pow(2) +
                cos(Math.toRadians(a.latitude)) * cos(Math.toRadians(b.latitude)) * sin(Math.toRadians(delta(b.longitude - a.longitude)) / 2).pow(2)
            lengths[i] = lengths[i - 1] + 6371008.8 * 2 * asin(sqrt(h.coerceIn(0.0, 1.0)))
        }
    }
    fun sample(progress: Double): MotionSample {
        val p = progress.coerceIn(0.0, 1.0)
        val rotation = if (separateRotation) min(1.0, p * 2) else p
        val movement = if (separateRotation) max(0.0, p * 2 - 1) else p
        val traveled = lengths.last() * movement
        var low = 1; var high = path.lastIndex
        while (low < high) { val mid = (low + high) / 2; if (lengths[mid] < traveled) low = mid + 1 else high = mid }
        val a = path[low - 1]; val b = path[low]
        val fraction = if (lengths[low] == lengths[low - 1]) movement else (traveled - lengths[low - 1]) / (lengths[low] - lengths[low - 1])
        val dx = Math.toRadians(delta(b.longitude - a.longitude))
        val bearing = Math.toDegrees(atan2(sin(dx) * cos(Math.toRadians(b.latitude)),
            cos(Math.toRadians(a.latitude)) * sin(Math.toRadians(b.latitude)) - sin(Math.toRadians(a.latitude)) * cos(Math.toRadians(b.latitude)) * cos(dx)))
        return MotionSample(MotionPoint(if (movement == 1.0) b.longitude else delta(a.longitude + delta(b.longitude - a.longitude) * fraction),
            a.latitude + (b.latitude - a.latitude) * fraction),
            if (autoRotate && movement > 0 && lengths.last() > 0) (bearing + 360) % 360 else angle + delta(rotate - angle) * rotation, traveled)
    }
    private fun delta(value: Double) = ((value + 540) % 360 + 360) % 360 - 180
}
